components/raftstore-v2/src/router/message.rs

```diff
@@ -0,0 +1,293 @@
+    pub snapshot: RegionSnapshot<S>,
+    // What is this?
+    pub txn_extra_op: TxnExtraOp,
+}
+
+impl<S: Snapshot> ReadResponseChannel<S> {
+    pub fn notify_read(&self, _res: Result<ReadResponse<S>, RaftCmdResponse>) {}
+}
+
+// This is only necessary because of seeming limitations in derive(Clone) w/r/t
+// generics. If it can be deleted in the future in favor of derive, it should
+// be.
+impl<S> Clone for ReadResponse<S>
+where
+    S: Snapshot,
+{
+    fn clone(&self) -> ReadResponse<S> {
+        ReadResponse {
+            snapshot: self.snapshot.clone(),
+            txn_extra_op: self.txn_extra_op,
+        }
+    }
+}
+
+/// Variants of channels for `Msg`.
+///  - `Read`: a channel for read only requests including `StatusRequest`,
+///         `GetRequest` and `SnapRequest`
+///  - `Write`: a channel for write only requests including `AdminRequest`
+///          `PutRequest`, `DeleteRequest` and `DeleteRangeRequest`.
+/// Prefer channel rather than callback because:
+/// 1. channel can be reused, hence reduce allocations.
+/// 2. channel may not need dynamic dispatch.
+/// 3. caller can use async fashion.
+/// 4. there will be no callback leak.
+pub enum ResponseChannel<S> {
+    /// No callback.
+    None,
+    /// Read callback.
+    Read(ReadResponseChannel<S>),
+    /// Write callback.
+    Write(WriteResponseChannel),
+}
+
+impl<S: Snapshot> HeapSize for ResponseChannel<S> {}
+
+impl<S> ResponseChannel<S>
+where
+    S: Snapshot,
+{
+    pub fn notify_applied(self, resp: RaftCmdResponse) {
+        match self {
+            ResponseChannel::None => (),
+            ResponseChannel::Read(read) => {
+                read.notify_read(Err(resp));
+            }
+            ResponseChannel::Write(write) => {
+                write.notify_applied(Err(resp));
+            }
+        }
+    }
+
+    pub fn notify_proposed(&mut self) {
+        if let ResponseChannel::Write(write) = self {
+            write.notify_proposed();
+        }
+    }
+
+    pub fn notify_committed(&mut self) {
+        if let ResponseChannel::Write(write) = self {
+            write.notify_committed();
+        }
+    }
+
+    pub fn invoke_read(self, args: ReadResponse<S>) {
+        match self {
+            ResponseChannel::Read(read) => read.notify_read(Ok(args)),
+            other => panic!("expect Callback::Read(..), got {:?}", other),
+        }
+    }
+
+    pub fn is_none(&self) -> bool {
+        matches!(self, ResponseChannel::None)
+    }
+}
+
+impl<S> fmt::Debug for ResponseChannel<S>
+where
+    S: Snapshot,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ResponseChannel::None => write!(fmt, "Callback::None"),
+            ResponseChannel::Read(_) => write!(fmt, "Callback::Read(..)"),
+            ResponseChannel::Write { .. } => write!(fmt, "Callback::Write(..)"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[repr(u8)]
+pub enum PeerTick {
```

Comment:
why not use V1's implementation as they're identical?

Reply:
Several ticks will be moved from `StoreTick` to `PeerTick`.